src/OpenSpaceToolkit/Astrodynamics/Trajectory/Segment.cpp

```diff
@@ -584,38 +622,269 @@ const NumericalSolver& Segment::accessNumericalSolver() const
+        nextSubSegmentToSolve = buildCoastSegment("(Coast - Minimum Maneuver Separation Constraint)");
+    }
+
+    // TBI: is this necessary?
+    const Duration subSegmentManeuverMargin = Duration::Seconds(10.0);
+
+    while (segmentStates.accessLast().accessInstant() < maximumInstant)
+    {
+        // Coast
+        if (nextSubSegmentToSolve.getType() == Segment::Type::Coast)
+        {
+            const Segment::Solution coastSegmentSolution = nextSubSegmentToSolve.solve(
+                segmentStates.accessLast(),
+                std::min(nextInstantToCoastTo, maximumInstant) - segmentStates.accessLast().accessInstant()
+            );
+            subSegmentConditionIsSatisfied = coastSegmentSolution.conditionIsSatisfied;
+
+            segmentStates.add(Array<State>(coastSegmentSolution.states.begin() + 1, coastSegmentSolution.states.end()));
+
+            if (subSegmentConditionIsSatisfied)
+            {
+                break;
+            }
+
+            nextSubSegmentToSolve = *this;
+
+            continue;
+        }
+
+        // Maneuver
+        const Segment::Solution maneuverSubsegmentSolution = nextSubSegmentToSolve.solve_(
+            segmentStates.accessLast(), maximumInstant - segmentStates.accessLast().accessInstant(), false
+        );
+        subSegmentConditionIsSatisfied = maneuverSubsegmentSolution.conditionIsSatisfied;
+
+        const Array<flightManeuver> subsegmentManeuvers =
+            maneuverSubsegmentSolution.extractManeuvers(aState.accessFrame());
+
+        // There were no maneuvers to be considered
+        if (subsegmentManeuvers.isEmpty())
+        {
+            segmentStates.add(
+                Array<State>(maneuverSubsegmentSolution.states.begin() + 1, maneuverSubsegmentSolution.states.end())
+            );
+            break;
+        }
+
+        // This should never happen, would indicate a problem with the solve_ method
+        if (subsegmentManeuvers.getSize() > 1)
+        {
+            throw ostk::core::error::RuntimeError(
+                "More than one maneuver found when solving maneuver segment with constraints."
+            );
+        }
+
+        const Interval candidateManeuverInterval = subsegmentManeuvers.accessFirst().getInterval();
+
+        // Check minimum maneuver duration constraint
+        if (maneuverConstraints_.minimumDuration.isDefined() &&
+            candidateManeuverInterval.getDuration() < maneuverConstraints_.minimumDuration)
+        {
+            BOOST_LOG_TRIVIAL(debug
+            ) << "Maneuver duration is less than the minimum maneuver duration. Skipping the maneuver."
+              << std::endl;
+
+            nextSubSegmentToSolve =
+                buildCoastSegment("(Skipping Maneuver due to Minimum Maneuver Duration Constraint)");
+
+            nextInstantToCoastTo =
+                updateNextCoastInstant(candidateManeuverInterval.getEnd() + subSegmentManeuverMargin);
```

Comment:
Is my understand correct that the `nextInstantToCoastTo` is actually more like a "next instant to coast to if the next subsegment to solve is a coast subsegment"? So its only used if the next subsegment turn out to be a subsegment that's either a coast segment from the start or a maneuver segment that violates a constraint and gets converted to coasting?

If both of those statements are correct, could you make that clearer with comments/variable naming?